xdsl/dialects/fsm.py

```diff
@@ -0,0 +1,379 @@
+                raise VerifyException(
+                    "The number of arg_attrs and arg_names should be the same"
+                )
+        if self.res_attrs is not None and self.res_names is not None:
+            if len(self.res_attrs) != len(self.res_names):
+                raise VerifyException(
+                    "The number of res_attrs and res_names should be the same"
+                )
+
+
+@irdl_op_definition
+class OutputOp(IRDLOperation):
+    """
+    Represents the outputs of a machine under a specific state. The
+    types of `$operands` should be consistent with the output types of the state
+    machine
+    """
+
+    name = "fsm.output"
+
+    operand = var_operand_def(AnyAttr())
+
+    traits = frozenset([IsTerminator()])
+
+    def __init__(
+        self,
+        operand: Sequence[SSAValue | Operation],
+    ):
+        super().__init__(
+            operands=[operand],
+        )
+
+    def verify_(self):
+        parent = self.parent_op()
+        if (
+            isinstance(parent, StateOp)
+            and parent.transitions == self.parent_region()
+            and len(self.operands) > 0
+        ):
+            raise VerifyException("Transition regions should not output any value")
+        while parent is not None:
+            if isinstance(parent, MachineOp):
+                if not (
+                    [operand.type for operand in self.operands]
+                    == [type(result) for result in parent.function_type.outputs]
+                    and len(self.operands) == len(parent.function_type.outputs)
+                ):
+                    raise VerifyException(
+                        "Output types must be consistent with the machine's"
+                    )
+            parent = parent.parent_op()
+
+
+@irdl_op_definition
+class StateOp(IRDLOperation):
+    """
+    Represents a state of a state machine. This op includes an
+    `$output` region with an `fsm.output` as terminator to define the machine
+    outputs under this state. This op also includes a `transitions` region that
+    contains all the transitions of this state
+    """
+
+    name = "fsm.state"
+
+    output = region_def()
+
+    transitions = region_def()
+
+    sym_name = attr_def(StringAttr)
+
+    traits = frozenset([NoTerminator(), SymbolOpInterface()])
+
+    def __init__(
+        self,
+        sym_name: str,
+        output: Region | type[Region.DEFAULT] = Region.DEFAULT,
+        transitions: Region | type[Region.DEFAULT] = Region.DEFAULT,
+    ):
+        attributes: dict[str, Attribute] = {
+            "sym_name": StringAttr(sym_name),
+        }
+        if not isinstance(output, Region):
+            output = Region(Block())
+        if not isinstance(transitions, Region):
+            transitions = Region(Block())
+        super().__init__(
+            attributes=attributes,
+            regions=[output, transitions],
+        )
+
+    def verify_(self):
+        parent = self.parent_op()
+
```

Comment:
I would just add an `assert(isinstance(parent, Machine))`, and remove the `while` loop.
It seems that even though they have the while loop in MLIR (through a function call), this is useless, since the parent is always the `Machine`.